src/libmongoc/tests/test-mongoc-dns.c

```diff
@@ -728,6 +728,117 @@ prose_test_9_pooled (void *unused)
    _prose_test_9 (true);
 }
 
+typedef struct {
+   int num_topology_opening;
+   int num_topology_closed;
+   int num_server_opening;
+   int num_server_closed;
+} cb_stats_t;
+
+static void
+invalid_topology_opening (const mongoc_apm_topology_opening_t *event) {
+   cb_stats_t *stats = (cb_stats_t*) mongoc_apm_topology_opening_get_context (event);
+   stats->num_topology_opening++;
+}
+
+static void
+invalid_topology_closed (const mongoc_apm_topology_closed_t *event) {
+   cb_stats_t *stats = (cb_stats_t*) mongoc_apm_topology_closed_get_context (event);
+   stats->num_topology_closed++;
+}
+
+static void
+invalid_server_closed (const mongoc_apm_server_closed_t *event) {
+   cb_stats_t *stats = (cb_stats_t*) mongoc_apm_server_closed_get_context (event);
+   stats->num_server_closed++;
+}
+
+static void
+invalid_server_opening (const mongoc_apm_server_opening_t *event) {
+   cb_stats_t *stats = (cb_stats_t*) mongoc_apm_server_opening_get_context (event);
+   stats->num_server_opening++;
+}
+
+/* CDRIVER-4184 Test that an invalid topology does not emit a topology_closed
+ * event. */
+static void
+test_invalid_topology_pooled (void *unused) {
+   mongoc_client_pool_t *pool;
+   mongoc_client_t *client;
+   mongoc_uri_t *uri;
+   mongoc_apm_callbacks_t *cbs;
+   cb_stats_t stats = {0};
+
+   /* TXT record for test20.test.build.10gen.cc resolves to "loadBalanced=true". */
+   uri = mongoc_uri_new ("mongodb+srv://test20.test.build.10gen.cc/?replicaSet=rs0");
```

Comment:
In many environments, a unit test against a resources like this is considered an invalid strategy. It's not clear to me that we're ok with this inside of 10gen-- that is, if this is a known-ok technique, so I'm pointing it out.

Reply:
Ok, @kmahar clarified this in her comment I'm with you now. +1.